### PR TITLE
STYLE add future annotations where possible

### LINKED ISSUE #41901: STYLE add check for future annotations
xref https://github.com/pandas-dev/pandas/pull/41892#issuecomment-857834651

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [x] closes #41901 (Replace xxxx with the Github issue number)
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -94,8 +94,6 @@ repos:
         stages: [manual]
         additional_dependencies: &pyright_dependencies
         - pyright@1.1.258
--   repo: local
-    hooks:
```
Comment (MarcoGorelli) [inline .pre-commit-config.yaml:98]: not relevant to this PR, but there's no need to keep repeating `repo: local`, so removing these extra ones while I'm here

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -237,3 +231,10 @@ repos:
         additional_dependencies:
         - flake8==4.0.1
         - flake8-pyi==22.5.1
+    -   id: future_annotations
+        name: import annotations from __future__
+        entry: 'from __future__ import annotations'
+        language: pygrep
+        args: [--negate]
```
Comment (MarcoGorelli) [inline .pre-commit-config.yaml:238]: `negate` checks that the pattern in `entry` is always present (as opposed to checking it's absent)

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -237,3 +231,15 @@ repos:
         additional_dependencies:
         - flake8==4.0.1
         - flake8-pyi==22.5.1
+    -   id: future-annotations
+        name: import annotations from __future__
+        entry: 'from __future__ import annotations'
+        language: pygrep
+        args: [--negate]
+        files: ^pandas/
+        types: [python]
```
Comment (twoertwein) [inline .pre-commit-config.yaml:240]: `python` excludes *.pyi files, or? pyi files wouldn't need this future import (https://github.com/PyCQA/flake8-pyi Y044)

Comment (MarcoGorelli) [reply .pre-commit-config.yaml:240]: yeah `python` is just Python files, `pyi` files have their own type (you can use `identify` to check):
```
$ identify-cli pandas/_libs/ops.pyi
["file", "non-executable", "pyi", "text"]
```

**Review by twoertwein — APPROVED** [review]

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @MarcoGorelli

**Pull request merged**